File: backend/src/jobs/retencaoJob.js
// src/jobs/retencaoJob.js
// Aplica a política de retenção periodicamente.
//
// Não usa cron: correr uma vez por dia com hora marcada obrigaria o contentor a
// estar acordado a essa hora, e no plano gratuito ele suspende sozinho. Um
// intervalo simples corre sempre que o serviço estiver de pé, e a política é
// idempotente — apagar o que já não existe não custa nada.
//
// É por isso que também não há recuperação de execuções perdidas: não há nada a
// recuperar. Se o serviço esteve em baixo três dias, a limpeza seguinte apanha
// os três dias de uma vez.
const logger = require('../config/logger');
const retencao = require('../services/retencaoService');

// Uma hora depois do arranque: o arranque já tem migrações, seed e a primeira
// vaga de pedidos: não é altura para uma limpeza.
const ESPERA_APOS_ARRANQUE_MS = 60 * 60 * 1000;
const INTERVALO_MS = 24 * 60 * 60 * 1000;

let emCurso = null;

async function correr() {
  // Single-flight, como nas cópias de segurança: dois DELETE grandes ao mesmo
  // tempo na mesma tabela é a receita para um lock desnecessário.
  if (emCurso) return emCurso;
  emCurso = (async () => {
    try {
      return await retencao.limpar();
    } catch (err) {
      // Falhar a limpeza não pode derrubar o serviço — é higiene, não é o
      // caminho crítico. Mas fica registado: uma limpeza que nunca corre é uma
      // política de retenção que não se cumpre.
      logger.error(`Retenção: a limpeza falhou — ${err.message}`);
      return null;
    } finally {
      emCurso = null;
    }
  })();
  return emCurso;
}

function scheduleRetencaoJob() {
  setTimeout(correr, ESPERA_APOS_ARRANQUE_MS).unref?.();
  setInterval(correr, INTERVALO_MS).unref?.();
  logger.info('Retenção de dados: limpeza agendada (diária)');
}

module.exports = { scheduleRetencaoJob, correr };

File: backend-java/src/main/java/ao/kixima/retention/RetencaoJob.java
package ao.kixima.retention;

import ao.kixima.ops.OperationalAlertService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

/**
 * Espelha backend/src/jobs/retencaoJob.js — aplica a política de retenção
 * periodicamente. Sem cron a hora fixa de propósito (mesma razão do Node:
 * um plano gratuito de hosting suspende sem tráfego; um atraso simples
 * corre sempre que o processo estiver de pé, e a política é idempotente).
 *
 * DESLIGADO POR OMISSÃO ({@code kixima.jobs.retencao.enabled: false}) —
 * o Node continua a correr esta limpeza em produção; ligar os dois ao
 * mesmo tempo duplicaria o trabalho (inofensivo aqui, por ser idempotente,
 * mas continua a ser a regra do plano para M6: nunca ligado por omissão
 * enquanto os dois backends coexistem).
 */
@Component
@ConditionalOnProperty(name = "kixima.jobs.retencao.enabled", havingValue = "true")
public class RetencaoJob {

    private static final Logger log = LoggerFactory.getLogger(RetencaoJob.class);

    private final RetentionService retentionService;
    private final OperationalAlertService operationalAlertService;

    public RetencaoJob(RetentionService retentionService, OperationalAlertService operationalAlertService) {
        this.retentionService = retentionService;
        this.operationalAlertService = operationalAlertService;
    }

    /** Uma hora depois do arranque (migrações/seed/primeira vaga de pedidos já passaram), depois a cada 24h. */
    @Scheduled(initialDelay = 60 * 60 * 1000, fixedDelay = 24 * 60 * 60 * 1000)
    public void correr() {
        try {
            retentionService.limpar();
        } catch (Exception e) {
            log.error("Retenção: a limpeza falhou — {}", e.getMessage());
            operationalAlertService.avisarFalha("RETENCAO_DADOS", "a limpeza de retenção de dados falhou",
                    "A limpeza automática não correu.\n\nErro: " + e.getMessage() + "\n\n"
                            + "A política de retenção publicada em /api/retencao promete apagar dados ao fim de um prazo. "
                            + "Enquanto isto falhar, a plataforma guarda mais do que diz guardar.");
        }
    }
}
